internal/pnp/pnp.go

```diff
@@ -0,0 +1,142 @@
+package pnp
+
+import (
+	"runtime"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"testing"
+)
+
+var (
+	isPnpApiInitialized atomic.Uint32
+	cachedPnpApi        *PnpApi
+	pnpMu               sync.Mutex
+	// testPnpCache stores per-goroutine PnP APIs for test isolation
+	// Key is goroutine ID (as int)
+	testPnpCache sync.Map // map[int]*PnpApi
+)
+
+// getGoroutineID returns the current goroutine ID
+// It is usually not recommended to work with goroutine IDs, but it is the most non-intrusive way to setup a parallel testing environment for PnP API
+func getGoroutineID() int {
+	var buf [64]byte
+	n := runtime.Stack(buf[:], false)
+	idField := strings.Fields(strings.TrimPrefix(string(buf[:n]), "goroutine "))[0]
+	id, _ := strconv.Atoi(idField)
+	return id
+}
+
+// Sets the PnP API for the given manifest data and manifest directory, used for testing
+// This creates a goroutine-specific cache entry that won't interfere with other parallel tests.
+func OverridePnpApi(fs PnpApiFS, manifestDataRaw string) *PnpApi {
+	if manifestDataRaw == "" {
+		return nil
+	}
+
+	var pnpApi *PnpApi
+
+	manifestData, err := parseManifestFromData(manifestDataRaw, "/")
+	if err != nil {
+		pnpApi = nil
+	} else if manifestData != nil {
+		pnpApi = &PnpApi{
+			fs:       fs,
+			url:      "/",
+			manifest: manifestData,
+		}
+	}
+
+	// Store in goroutine-specific cache using goroutine ID
+	// This allows parallel tests to have isolated PnP configurations
+	gid := getGoroutineID()
+	testPnpCache.Store(gid, pnpApi)
+
+	return pnpApi
+}
+
+// ClearTestPnpCache clears the test-specific PnP API cache for the current goroutine
+func ClearTestPnpCache() {
+	gid := getGoroutineID()
+	testPnpCache.Delete(gid)
+}
+
+func InitPnpApi(fs PnpApiFS, filePath string) *PnpApi {
+	pnpMu.Lock()
+	defer pnpMu.Unlock()
+	// Double-check after acquiring lock
+	if isPnpApiInitialized.Load() == 1 {
+		return cachedPnpApi
+	}
+
+	pnpApi := &PnpApi{fs: fs, url: filePath}
+
+	manifestData, err := pnpApi.findClosestPnpManifest()
```

Comment:
I don't really understand this init. Surely one needs to be able to load multiple projects with differing pnp info at the same time? All of this info really does need to be handled differently.